### PR TITLE
docs: add athens protocol adr

### PR DESCRIPTION
**Pull request by filipesilva**

### CODE REVIEW

File: doc/adr/0016-athens-protocol.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+Client nodes can create a new optimistic state by applying optimistic operations over the last known state.
+Synchronization can happen as frequently or infrequently as desired.
+Each operation a node creates has a unique id used for identification and to prevent duplication in the log.
+
+Standalone clients do not need to synchronize with other nodes, but still express changes in the same format.
+They are, essentially, a client node that never synchronizes with other nodes.
+
+Different nodes can arrive at different states based on their interpretation of the operations.
+This way nodes can select for information that is relevant to them.
+This also allows for different node capabilities and conflict resolution strategies.
```
Comment (neotyk) [inline doc/adr/0016-athens-protocol.md:45]: I find it too generic for my liking.
What I think you're trying to say is something like "Capture only mobile client" or similar reduced functionality client, right?
If so an example or two would help a ton.

File: doc/adr/0016-athens-protocol.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+The initial set of atomic operations expresses only a tree structure via page identity and block identity, content and location.
+For now, content parsing for references is not part of the protocol itself, and is instead left to the nodes.
```
Comment (neotyk) [inline doc/adr/0016-athens-protocol.md:48]: Reference ADRs 10 & 14?

File: doc/adr/0016-athens-protocol.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+
+## Consequences
+
+The canonical log requires nodes to achieve consensus on what the order of operations is.
+
+The latest graph state can only be obtained by replaying the full log, or via querying for a node-compatible view of the current state.
```
Comment (neotyk) [inline doc/adr/0016-athens-protocol.md:55]: Or reading an Internal Representation dump file?

**Review by neotyk — CHANGES_REQUESTED** [review]

File: doc/adr/0016-athens-protocol.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+
+## Status
+
+Accepted.
+
+
+## Context
+
+The Athens Protocol defines the format and semantics for messages between Athens nodes.
```
[the remark this answers is not in the document]
Comment (filipesilva) [reply doc/adr/0016-athens-protocol.md]: Ok, but this doc is mostly about the protocol, its properties and motivation, not so much about the current implementation we have.

File: doc/adr/0016-athens-protocol.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+The protocol aims to support a few key requirements:
+  1. longevity: information must be usable for at least a long period of time in the future, e.g. 10+ years
+  2. robustness: nodes must be able to synchronize through unreliable network connections
+  3. reliability: data must not be lost and remain acessible
+  4. extensibility: the protocol must support extension over time
+
+
+## Decision
+
+We decided to model the Athens Protocol as an append-only log of deterministic operations.
```
[the remark this answers is not in the document]
Comment (filipesilva) [reply doc/adr/0016-athens-protocol.md]: Our currently RTC clients make use of a server to obtain consensus on the ordering of events/operations, and to obtains views (the internal representation) of a state. The key part here is not that we have a server, but what the server is doing.

Hm the distinction between event and operation is not very clear here... The word event isn't anywhere else in this doc. And the log definitely lists operations. But like this it sounds like each different atomic/composite op has a unique id, which isn't true.

I don't think events as we have them currently are a concept in the protocol. We happen to wrap both atomic/composite operations and presence events in a more general event, but it's just a wrapper. The concept of event is not important for this document, the important concept is that of an ordered log of operations. I'd rather not introduce another concept here that isn't necessary.

File: doc/adr/0016-athens-protocol.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+Operations from multiple nodes are weaved together into a single canonical log.
```
[the remark this answers is not in the document]
Comment (filipesilva) [reply doc/adr/0016-athens-protocol.md]: Done

File: doc/adr/0016-athens-protocol.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+Knowledge graph state is determined by the reduction of all operations.
+The current state can be represented by a canonical view.
+
+Operations express semantic changes to the knowledge graph.
+Non-trivial operations are expressed as composites of atomic operations.
```
[the remark this answers is not in the document]
Comment (filipesilva) [reply doc/adr/0016-athens-protocol.md]: Done

File: doc/adr/0016-athens-protocol.md [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+The effect of atomic operations is kept small and local to their direct relationships.
+Future versions of the protocol can enhance the vocabulary of atomic operations, but can never remove existing ones.
```
[the remark this answers is not in the document]
Comment (filipesilva) [reply doc/adr/0016-athens-protocol.md]: Added, also added that relaxing of constraints is ok but tightening is not.

File: doc/adr/0016-athens-protocol.md [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+
+## Decision
+
+We decided to model the Athens Protocol as an append-only log of deterministic operations.
+Operations from multiple nodes are weaved together into a single canonical log.
+Knowledge graph state is determined by the reduction of all operations.
+The current state can be represented by a canonical view.
+
+Operations express semantic changes to the knowledge graph.
```
[the remark this answers is not in the document]
Comment (filipesilva) [reply doc/adr/0016-athens-protocol.md]: For the reasons I left in in the other comment, I'd rather not talk about events.

File: doc/adr/0016-athens-protocol.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+Non-trivial operations are expressed as composites of atomic operations.
+The effect of atomic operations is kept small and local to their direct relationships.
+Future versions of the protocol can enhance the vocabulary of atomic operations, but can never remove existing ones.
+
+Client nodes synchronize state by sending optimistic operations, and listening to the canonical stream of operations from the server.
+Client nodes can create a new optimistic state by applying optimistic operations over the last known state.
+Synchronization can happen as frequently or infrequently as desired.
+Each operation a node creates has a unique id used for identification and to prevent duplication in the log.
```
[the remark this answers is not in the document]
Comment (filipesilva) [reply doc/adr/0016-athens-protocol.md]: For the reasons I left in in the other comment, I'd rather not talk about events. WDYT of "Each synchronized operation has a unique id ..."?

File: doc/adr/0016-athens-protocol.md [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+Standalone clients do not need to synchronize with other nodes, but still express changes in the same format.
```
[the remark this answers is not in the document]
Comment (filipesilva) [reply doc/adr/0016-athens-protocol.md]: All of that's true, but what would you like me to change here?

File: doc/adr/0016-athens-protocol.md (hunk 1 above)
Comment (filipesilva) [reply doc/adr/0016-athens-protocol.md]: Yeah examples here would really help, added.

File: doc/adr/0016-athens-protocol.md (hunk 2 above)
Comment (filipesilva) [reply doc/adr/0016-athens-protocol.md]: Just noticed we have two ADR 10 :/

Added ADR 10 ref, and renamed the other ADR 10 to 16 (and this ADR to 17).

I don't follow where ADR 14 should be reffed.  ADR 14 is about resolve-transact iteratively, which also doesn't sound related.

File: doc/adr/0016-athens-protocol.md (hunk 3 above)
Comment (filipesilva) [reply doc/adr/0016-athens-protocol.md]: Internal representation is a view of the current state. That it is in a file is not important. DB-dump right now is a good example of this mechanism.

**Comment by filipesilva** [comment]
Also changes `nodes` to `clients` following our convo on discord.

**Review by neotyk — APPROVED** [review]

**Pull request merged**